### PR TITLE
Add deferrable ``GCSObjectExistenceSensorAsync``

### PR DESCRIPTION
**Pull request by rajaths010494**
This PR donates the following developed GCSObjectExistenceSensorAsync` in [astronomer-providers](https://github.com/astronomer/astronomer-providers) repo to apache airflow.

`GCSObjectExistenceSensorAsync`

### CODE REVIEW

File: airflow/providers/google/cloud/hooks/gcs.py [hunk 1]
```diff
@@ -32,18 +32,21 @@
 from typing import IO, Callable, Generator, Sequence, TypeVar, cast, overload
 from urllib.parse import urlsplit
 
+from aiohttp import ClientSession as ClientSession
```
Comment (uranusjr) [inline airflow/providers/google/cloud/hooks/gcs.py]:
```suggestion
from aiohttp import ClientSession
```

Comment (rajaths010494) [reply airflow/providers/google/cloud/hooks/gcs.py]: changed.

**Comment by rajaths010494** [comment]
When i re-requested for review. It automatically removed reviewers and now i am not able to add them back.

**Review by phanikumv — APPROVED** [review]

File: airflow/providers/google/cloud/sensors/gcs.py [hunk 2]
```diff
@@ -94,6 +95,55 @@ def poke(self, context: Context) -> bool:
...
+                google_cloud_conn_id=self.google_cloud_conn_id,
+                hook_params={
+                    "delegate_to": self.delegate_to,
+                    "impersonation_chain": self.impersonation_chain,
+                },
+            ),
+            method_name="execute_complete",
+        )
+
+    def execute_complete(self, context: dict[str, Any], event: dict[str, str]) -> str:
```
Comment (ashb) [inline airflow/providers/google/cloud/sensors/gcs.py]: I think?
```suggestion
    def execute_complete(self, context: Context, event: dict[str, str]) -> str:
```

File: airflow/providers/google/cloud/hooks/gcs.py [hunk 3]
```diff
@@ -1174,3 +1177,14 @@ def _parse_gcs_url(gsurl: str) -> tuple[str, str]:
     # Remove leading '/' but NOT trailing one
     blob = parsed_url.path.lstrip("/")
     return bucket, blob
+
+
+class GCSAsyncHook(GoogleBaseAsyncHook):
+    """GCSAsyncHook run on the trigger worker, inherits from GoogleBaseHookAsync"""
+
+    sync_hook_class = GCSHook
```
Comment (utkarsharma2) [inline airflow/providers/google/cloud/hooks/gcs.py:1185]: @phanikumv If this is a constant should be in caps - `SYNC_HOOK_CLASS`

File: airflow/providers/google/cloud/sensors/gcs.py (hunk 2 above)
Comment (rajaths010494) [reply airflow/providers/google/cloud/sensors/gcs.py]: Modified

**Review by kaxil — APPROVED** [review]

**Comment by vchiapaikeo** [comment]
Hi @rajaths010494 @kaxil - I wanted to point out a possible bug that is introduced here - it seems like using this approach:

```py
    async def get_storage_client(self, session: ClientSession) -> Storage:
        """Returns a Google Cloud Storage service object."""
        with await self.service_file_as_context() as file:
            return Storage(service_file=file, session=cast(Session, session))
```

does not honor the impersonation_chain when application default credentials (ADC) are being used. We've tried a similar approach from astronomer-providers and encountered issues on our end. I believe this is because in the case that `await self.service_file_as_context() as file` yields None (the fall through [here](https://github.com/apache/airflow/blob/284cd529898fbadd14308004a0b0cb6f389b4318/airflow/providers/google/common/hooks/base_google.py#L484-L513)), the async storage client gets instantiated with this:

```
return Storage(service_file=None, session=cast(Session, session))
```

TalkIQ's async io storage client has no idea of the impersonation chain on instantiation here --> https://github.com/talkiq/gcloud-aio/blob/4e0006240913ecadde68722b8986f184a9b1adea/storage/gcloud/aio/storage/storage.py#L149-L168

And as a result, the token gets instantiated with ADC without regard for impersonation chain.

**Comment by rajaths010494** [comment]
Thanks @vchiapaikeo i will look at this and check [get_credentials_and_project_id](https://github.com/apache/airflow/blob/284cd529898fbadd14308004a0b0cb6f389b4318/airflow/providers/google/common/hooks/base_google.py#L235-L271) which handles the impersonation chain properly.
and add  support for impersonation chain

**Pull request merged**